lib/model/model.go

```diff
@@ -2432,42 +2441,33 @@ func (m *model) GlobalDirectoryTree(folder, prefix string, levels int, dirsonly
 
 		f.Name = strings.Replace(f.Name, prefix, "", 1)
 
-		var dir, base string
-		if f.IsDirectory() && !f.IsSymlink() {
-			dir = f.Name
-		} else {
-			dir = filepath.Dir(f.Name)
-			base = filepath.Base(f.Name)
-		}
-
 		if levels > -1 && strings.Count(f.Name, sep) > levels {
 			return true
 		}
 
-		last := output
-		if dir != "." {
-			for _, path := range strings.Split(dir, sep) {
-				directory, ok := last[path]
-				if !ok {
-					newdir := make(map[string]interface{})
-					last[path] = newdir
-					last = newdir
-				} else {
-					last = directory.(map[string]interface{})
-				}
-			}
+		fullName := strings.Split(f.Name, sep)
+		name := fullName[len(fullName)-1]
+		path := fullName[:len(fullName)-1]
+
+		tree := DirectoryTree{Name: name, IsDirectory: f.IsDirectory(), Children: []*DirectoryTree{}}
+		for !reflect.DeepEqual(path, currentPrefix) && len(currentPrefix) != 0 {
+			breadcrumbs = breadcrumbs[:len(breadcrumbs)-1]
+			currentPrefix = currentPrefix[:len(currentPrefix)-1]
 		}
 
-		if !dirsonly && base != "" {
-			last[base] = []interface{}{
-				f.ModTime(), f.FileSize(),
-			}
+		if !dirsonly || f.IsDirectory() {
+			breadcrumbs[len(breadcrumbs)-1].Children = append(breadcrumbs[len(breadcrumbs)-1].Children, &tree)
```

Comment:
I wonder if it's not easier to track "parent", and update it as we traverse deeper or traverse out.
I know array indexing is fast, but the code is a lot less readable.

Reply:
In my opinion, that would be the same amount of work, just in a different place in the code. But if you insist, I can try to implement it as you suggest.

Reply:
Okay, the idea with these 2 lines you wrote, seems more reasonable to me than tracking a parent. I don't have to rework the algorithm from scratch. I just split the oneliner.

Reply:
Splitting this line as proposed by Audrius would definitely help. And I just spent quite a bit of time figuring out how the entire loop works (if I did figure it out). Is this correct: `currentPrefix` contains the path components of the latest encountered directory and `breadcrumbs` the tree nodes aka `FolderItem`s corresponding to those path components? If that's correct, comments stating something like that when declaring those variables would be a good idea.

Reply:
@imsodin Yes, you figured it out correctly. The comments make sense.